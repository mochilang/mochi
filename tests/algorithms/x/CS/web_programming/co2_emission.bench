{"name":"main","duration_us":168557,"memory_bytes":2115432}