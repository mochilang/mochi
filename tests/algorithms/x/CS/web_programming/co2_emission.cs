// Generated by Mochi 0.10.63 on 2025-08-11 15:03 +0700
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Net.Http;
using System.Numerics;
using System.Collections;

class Intensity {
    public long forecast;
    public long actual;
    public string index;
    public override string ToString() => $"Intensity {{forecast = {forecast}, actual = {actual}, index = \"{index}\"}}";
}
class Entry {
    public string from;
    public string to;
    public Intensity intensity;
    public override string ToString() => $"Entry {{from = \"{from}\", to = \"{to}\", intensity = {intensity}}}";
}
class Response {
    public Entry[] data;
    public override string ToString() => $"Response {{data = {data}}}";
}
#pragma warning disable CS0162
class Program {
    static bool seededNow = false;
    static long nowSeed = 0;
    static long _now() {
        if (!seededNow) {
            var s = Environment.GetEnvironmentVariable("MOCHI_NOW_SEED");
            if (long.TryParse(s, out var v)) {
                nowSeed = v;
                seededNow = true;
            }
        }
        if (seededNow) {
            nowSeed = unchecked(nowSeed * 1664525 + 1013904223);
            nowSeed %= 9223372036854775783L;
            if (nowSeed < 0) nowSeed += 9223372036854775783L;
            return nowSeed;
        }
        return DateTime.UtcNow.Ticks / 100;
    }
    static long _mem() {
        return GC.GetTotalAllocatedBytes(true);
    }
    static long _len(object v) {
        if (v is Array a) return a.Length;
        if (v is string s) return s.Length;
        if (v is System.Collections.ICollection c) return c.Count;
        return Convert.ToString(v).Length;
    }
    static T _idx<T>(T[] arr, long i) {
        if (arr == null) return default(T);
        if (i < 0) i += arr.Length;
        if (i < 0 || i >= arr.Length) return default(T);
        return arr[(int)i];
    }
    static object _fetch(string url) {
        if (url.StartsWith("http://") || url.StartsWith("https://")) {
            using var client = new HttpClient();
            var resp = client.GetAsync(url).Result;
            using var stream = resp.Content.ReadAsStream();
            var opts = new JsonSerializerOptions{ IncludeFields = true };
            return JsonSerializer.Deserialize<object>(stream, opts);
        } else {
            var path = url;
            if (!System.IO.Path.IsPathRooted(path)) {
                var root = Environment.GetEnvironmentVariable("MOCHI_ROOT");
                if (!string.IsNullOrEmpty(root)) {
                    var combined = System.IO.Path.Combine(root, path.Replace('/', System.IO.Path.DirectorySeparatorChar));
                    if (System.IO.File.Exists(combined)) path = combined;
                }
            }
            using var stream = System.IO.File.OpenRead(path);
            var opts = new JsonSerializerOptions{ IncludeFields = true };
            return JsonSerializer.Deserialize<object>(stream, opts);
        }
    }
    static string _fetch(string url, object optsObj) {
        if (!(optsObj is System.Collections.IDictionary o)) return Convert.ToString(_fetch(url));
        var tmp = new Dictionary<string, object>();
        foreach (System.Collections.DictionaryEntry kv in o) {
            var k = kv.Key.ToString();
            var i = k.IndexOf('_');
            if (i >= 0) k = k.Substring(0, i);
            tmp[k] = kv.Value;
        }
        var method = tmp.ContainsKey("method") ? Convert.ToString(tmp["method"]) : "GET";
        Dictionary<string, string>? headers = null;
        if (tmp.ContainsKey("headers")) {
            headers = new Dictionary<string, string>();
            if (tmp["headers"] is System.Collections.IDictionary hd) {
                foreach (System.Collections.DictionaryEntry kv in hd) {
                    headers[kv.Key.ToString()] = kv.Value.ToString();
                }
            }
        }
        object bodyObj = null;
        if (tmp.ContainsKey("body")) bodyObj = tmp["body"];
        if (url.StartsWith("http://") || url.StartsWith("https://")) {
            using var client = new HttpClient();
            if (method == "POST") {
                var jsonBody = bodyObj == null ? "" : JsonSerializer.Serialize(bodyObj);
                var content = new StringContent(jsonBody);
                if (headers != null) foreach (var kv in headers) content.Headers.TryAddWithoutValidation(kv.Key, kv.Value);
                var respPost = client.PostAsync(url, content).Result;
                return respPost.Content.ReadAsStringAsync().Result;
            }
            if (headers != null) foreach (var kv in headers) client.DefaultRequestHeaders.TryAddWithoutValidation(kv.Key, kv.Value);
            var resp = client.GetAsync(url).Result;
            return resp.Content.ReadAsStringAsync().Result;
        }
        return Convert.ToString(_fetch(url));
    }
    static Response _fetch_Response(string url) {
        if (url.StartsWith("http://") || url.StartsWith("https://")) {
            using var client = new HttpClient();
            var resp = client.GetAsync(url).Result;
            using var stream = resp.Content.ReadAsStream();
            var opts = new JsonSerializerOptions{ IncludeFields = true };
            var res = JsonSerializer.Deserialize<Response>(stream, opts);
            return res == null ? new Response() : res;
        } else {
            var path = url;
            if (!System.IO.Path.IsPathRooted(path)) {
                var root = Environment.GetEnvironmentVariable("MOCHI_ROOT");
                if (!string.IsNullOrEmpty(root)) {
                    var combined = System.IO.Path.Combine(root, path.Replace('/', System.IO.Path.DirectorySeparatorChar));
                    if (System.IO.File.Exists(combined)) path = combined;
                }
            }
            using var stream = System.IO.File.OpenRead(path);
            var opts = new JsonSerializerOptions{ IncludeFields = true };
            var res = JsonSerializer.Deserialize<Response>(stream, opts);
            return res == null ? new Response() : res;
        }
    }
    static Response _fetch_Response(string url, object optsObj) {
        var str = _fetch(url, optsObj);
        var opts = new JsonSerializerOptions{ IncludeFields = true };
        var res = JsonSerializer.Deserialize<Response>(str, opts);
        return res == null ? new Response() : res;
    }
    static string _substr(string s, long start, long end) {
        if (start < 0) start = 0;
        if (end < 0) end = 0;
        if (start > s.Length) start = s.Length;
        if (end > s.Length) end = s.Length;
        if (start > end) start = end;
        return s.Substring((int)start, (int)(end - start));
    }
    static string _fmt(object v) {
        if (v is Array a) {
            var parts = new List<string>();
            foreach (var x in a) parts.Add(_fmt(x));
            return "[" + string.Join(", ", parts) + "]";
        }
        if (v is System.Collections.IDictionary d) {
            var keys = new List<string>();
            foreach (var k in d.Keys) keys.Add(k.ToString());
            keys.Sort();
            var parts = new List<string>();
            foreach (var k in keys) parts.Add(k + ":" + _fmt(d[k]));
            return "map[" + string.Join(", ", parts) + "]";
        }
        if (v is System.Collections.IEnumerable e && !(v is string)) {
            var parts = new List<string>();
            foreach (var x in e) parts.Add(_fmt(x));
            return string.Join(", ", parts);
        }
        if (v is string s) return "\"" + s.Replace("\"", "\\\"") + "\"";
        if (v is bool b) return b ? "true" : "false";
        return Convert.ToString(v);
    }
    static string _fmtStr(object v) {
        if (v is Array a) {
            var parts = new List<string>();
            foreach (var x in a) parts.Add(_fmtStr(x));
            return "[" + string.Join(" ", parts) + "]";
        }
        if (v is System.Collections.IDictionary d) {
            var keys = new List<string>();
            foreach (var k in d.Keys) keys.Add(k.ToString());
            keys.Sort();
            var parts = new List<string>();
            foreach (var k in keys) parts.Add(k + ":" + _fmtStr(d[k]));
            return "map[" + string.Join(" ", parts) + "]";
        }
        if (v is System.Collections.IEnumerable e && !(v is string)) {
            var parts = new List<string>();
            foreach (var x in e) parts.Add(_fmtStr(x));
            return string.Join(" ", parts);
        }
        if (v is string s) return "\"" + s.Replace("\"", "\\\"") + "\"";
        if (v is bool b) return b ? "true" : "false";
        return Convert.ToString(v);
    }
    public static string _fmtTop(object v) {
        if (v is Array a && a.Length > 0 && a.GetValue(0) is Array) {
            var parts = new List<string>();
            foreach (var x in a) parts.Add(_fmt(x));
            return string.Join(" ", parts);
        }
        if (v is string s) return s;
        return _fmt(v);
    }
    static string __name__ = "__main__";
    static string BASE_URL_0 = "https://api.carbonintensity.org.uk/intensity";
    public static long fetch_last_half_hour() {
        Response resp_1 = _fetch_Response(BASE_URL_0, new Dictionary<string, double>{{"timeout", 10.0}});
        Entry entry_2 = _idx(resp_1.data, 0);
        return entry_2.intensity.actual;
    }

    public static Entry[] fetch_from_to(string start_3, string end_4) {
        string url_5 = ((((BASE_URL_0 + "/") + start_3) + "/") + end_4);
        Response resp_6 = _fetch_Response(url_5, new Dictionary<string, double>{{"timeout", 10.0}});
        return resp_6.data;
    }

    public static void main() {
        Entry[] entries_7 = Program.fetch_from_to("2020-10-01", "2020-10-03");
        long i_8 = 0;
        while ((i_8 < entries_7.Length)) {
            Entry e_9 = _idx(entries_7, i_8);
            Console.WriteLine(string.Join(" ", new string[]{Program._fmtTop("from"), Program._fmtTop(e_9.from), Program._fmtTop("to"), Program._fmtTop(e_9.to), Program._fmtTop(":"), Program._fmtTop(e_9.intensity.actual)}));
            i_8 = (i_8 + 1);
        };
        long last_10 = Program.fetch_last_half_hour();
        Console.WriteLine(string.Join(" ", new string[]{Program._fmtTop("fetch_last_half_hour() ="), Program._fmtTop(last_10)}));
    }

    static void Main() {
        {
            var __memStart = _mem();
            var __start = _now();
            Program.main();
            var __end = _now();
            var __memEnd = _mem();
            var __dur = (__end - __start);
            if (__dur <= 0) __dur = 1;
            var __memDiff = __memEnd - __memStart;
            if (__memDiff <= 0) __memDiff = __memEnd;
            Console.WriteLine("{\"name\":\"main\",\"duration_us\":" + __dur + ",\"memory_bytes\":" + __memDiff + "}");
        }
    }
}
