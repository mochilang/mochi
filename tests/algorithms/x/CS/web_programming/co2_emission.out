/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(92,35): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' annotations context. [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(23,20): warning CS0649: Field 'Response.data' is never assigned to, and will always have its default value null [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(13,19): warning CS0649: Field 'Intensity.index' is never assigned to, and will always have its default value null [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(11,17): warning CS0649: Field 'Intensity.forecast' is never assigned to, and will always have its default value 0 [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(19,22): warning CS0649: Field 'Entry.intensity' is never assigned to, and will always have its default value null [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(17,19): warning CS0649: Field 'Entry.from' is never assigned to, and will always have its default value null [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(18,19): warning CS0649: Field 'Entry.to' is never assigned to, and will always have its default value null [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(12,17): warning CS0649: Field 'Intensity.actual' is never assigned to, and will always have its default value 0 [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/Program.cs(210,19): warning CS0414: The field 'Program.__name__' is assigned but its value is never used [/tmp/TestCSTranspiler_Algorithms_Golden001_web_programming_co2_emission1723049593/001/app.csproj]
from 2020-09-30T23:30Z to 2020-10-01T00:00Z : 160
from 2020-10-01T00:00Z to 2020-10-01T00:30Z : 164
from 2020-10-01T00:30Z to 2020-10-01T01:00Z : 163
from 2020-10-01T01:00Z to 2020-10-01T01:30Z : 165
from 2020-10-01T01:30Z to 2020-10-01T02:00Z : 167
from 2020-10-01T02:00Z to 2020-10-01T02:30Z : 169
from 2020-10-01T02:30Z to 2020-10-01T03:00Z : 166
from 2020-10-01T03:00Z to 2020-10-01T03:30Z : 174
from 2020-10-01T03:30Z to 2020-10-01T04:00Z : 188
from 2020-10-01T04:00Z to 2020-10-01T04:30Z : 217
from 2020-10-01T04:30Z to 2020-10-01T05:00Z : 239
from 2020-10-01T05:00Z to 2020-10-01T05:30Z : 262
from 2020-10-01T05:30Z to 2020-10-01T06:00Z : 275
from 2020-10-01T06:00Z to 2020-10-01T06:30Z : 289
from 2020-10-01T06:30Z to 2020-10-01T07:00Z : 284
from 2020-10-01T07:00Z to 2020-10-01T07:30Z : 276
from 2020-10-01T07:30Z to 2020-10-01T08:00Z : 272
from 2020-10-01T08:00Z to 2020-10-01T08:30Z : 267
from 2020-10-01T08:30Z to 2020-10-01T09:00Z : 266
from 2020-10-01T09:00Z to 2020-10-01T09:30Z : 260
from 2020-10-01T09:30Z to 2020-10-01T10:00Z : 254
from 2020-10-01T10:00Z to 2020-10-01T10:30Z : 250
from 2020-10-01T10:30Z to 2020-10-01T11:00Z : 247
from 2020-10-01T11:00Z to 2020-10-01T11:30Z : 242
from 2020-10-01T11:30Z to 2020-10-01T12:00Z : 242
from 2020-10-01T12:00Z to 2020-10-01T12:30Z : 239
from 2020-10-01T12:30Z to 2020-10-01T13:00Z : 236
from 2020-10-01T13:00Z to 2020-10-01T13:30Z : 236
from 2020-10-01T13:30Z to 2020-10-01T14:00Z : 250
from 2020-10-01T14:00Z to 2020-10-01T14:30Z : 268
from 2020-10-01T14:30Z to 2020-10-01T15:00Z : 277
from 2020-10-01T15:00Z to 2020-10-01T15:30Z : 291
from 2020-10-01T15:30Z to 2020-10-01T16:00Z : 302
from 2020-10-01T16:00Z to 2020-10-01T16:30Z : 306
from 2020-10-01T16:30Z to 2020-10-01T17:00Z : 305
from 2020-10-01T17:00Z to 2020-10-01T17:30Z : 307
from 2020-10-01T17:30Z to 2020-10-01T18:00Z : 308
from 2020-10-01T18:00Z to 2020-10-01T18:30Z : 308
from 2020-10-01T18:30Z to 2020-10-01T19:00Z : 303
from 2020-10-01T19:00Z to 2020-10-01T19:30Z : 294
from 2020-10-01T19:30Z to 2020-10-01T20:00Z : 283
from 2020-10-01T20:00Z to 2020-10-01T20:30Z : 267
from 2020-10-01T20:30Z to 2020-10-01T21:00Z : 262
from 2020-10-01T21:00Z to 2020-10-01T21:30Z : 229
from 2020-10-01T21:30Z to 2020-10-01T22:00Z : 209
from 2020-10-01T22:00Z to 2020-10-01T22:30Z : 199
from 2020-10-01T22:30Z to 2020-10-01T23:00Z : 189
from 2020-10-01T23:00Z to 2020-10-01T23:30Z : 175
from 2020-10-01T23:30Z to 2020-10-02T00:00Z : 163
from 2020-10-02T00:00Z to 2020-10-02T00:30Z : 160
from 2020-10-02T00:30Z to 2020-10-02T01:00Z : 160
from 2020-10-02T01:00Z to 2020-10-02T01:30Z : 150
from 2020-10-02T01:30Z to 2020-10-02T02:00Z : 147
from 2020-10-02T02:00Z to 2020-10-02T02:30Z : 141
from 2020-10-02T02:30Z to 2020-10-02T03:00Z : 154
from 2020-10-02T03:00Z to 2020-10-02T03:30Z : 147
from 2020-10-02T03:30Z to 2020-10-02T04:00Z : 154
from 2020-10-02T04:00Z to 2020-10-02T04:30Z : 160
from 2020-10-02T04:30Z to 2020-10-02T05:00Z : 166
from 2020-10-02T05:00Z to 2020-10-02T05:30Z : 182
from 2020-10-02T05:30Z to 2020-10-02T06:00Z : 191
from 2020-10-02T06:00Z to 2020-10-02T06:30Z : 193
from 2020-10-02T06:30Z to 2020-10-02T07:00Z : 186
from 2020-10-02T07:00Z to 2020-10-02T07:30Z : 180
from 2020-10-02T07:30Z to 2020-10-02T08:00Z : 178
from 2020-10-02T08:00Z to 2020-10-02T08:30Z : 181
from 2020-10-02T08:30Z to 2020-10-02T09:00Z : 178
from 2020-10-02T09:00Z to 2020-10-02T09:30Z : 177
from 2020-10-02T09:30Z to 2020-10-02T10:00Z : 176
from 2020-10-02T10:00Z to 2020-10-02T10:30Z : 166
from 2020-10-02T10:30Z to 2020-10-02T11:00Z : 169
from 2020-10-02T11:00Z to 2020-10-02T11:30Z : 171
from 2020-10-02T11:30Z to 2020-10-02T12:00Z : 172
from 2020-10-02T12:00Z to 2020-10-02T12:30Z : 170
from 2020-10-02T12:30Z to 2020-10-02T13:00Z : 171
from 2020-10-02T13:00Z to 2020-10-02T13:30Z : 170
from 2020-10-02T13:30Z to 2020-10-02T14:00Z : 162
from 2020-10-02T14:00Z to 2020-10-02T14:30Z : 165
from 2020-10-02T14:30Z to 2020-10-02T15:00Z : 168
from 2020-10-02T15:00Z to 2020-10-02T15:30Z : 176
from 2020-10-02T15:30Z to 2020-10-02T16:00Z : 181
from 2020-10-02T16:00Z to 2020-10-02T16:30Z : 183
from 2020-10-02T16:30Z to 2020-10-02T17:00Z : 177
from 2020-10-02T17:00Z to 2020-10-02T17:30Z : 176
from 2020-10-02T17:30Z to 2020-10-02T18:00Z : 175
from 2020-10-02T18:00Z to 2020-10-02T18:30Z : 172
from 2020-10-02T18:30Z to 2020-10-02T19:00Z : 171
from 2020-10-02T19:00Z to 2020-10-02T19:30Z : 162
from 2020-10-02T19:30Z to 2020-10-02T20:00Z : 152
from 2020-10-02T20:00Z to 2020-10-02T20:30Z : 146
from 2020-10-02T20:30Z to 2020-10-02T21:00Z : 162
from 2020-10-02T21:00Z to 2020-10-02T21:30Z : 135
from 2020-10-02T21:30Z to 2020-10-02T22:00Z : 124
from 2020-10-02T22:00Z to 2020-10-02T22:30Z : 110
from 2020-10-02T22:30Z to 2020-10-02T23:00Z : 104
from 2020-10-02T23:00Z to 2020-10-02T23:30Z : 97
from 2020-10-02T23:30Z to 2020-10-03T00:00Z : 102
fetch_last_half_hour() = 139